source/opt/loop_descriptor.cpp

```diff
@@ -153,5 +250,122 @@ void LoopDescriptor::PopulateList(const Function* f) {
   }
 }
 
+ir::BasicBlock* Loop::FindConditionBlock(const ir::Function& function) {
+  ir::BasicBlock* condition_block = nullptr;
+
+  opt::DominatorAnalysis* dom_analysis =
+      ir_context_->GetDominatorAnalysis(&function, *ir_context_->cfg());
+  ir::BasicBlock* bb = dom_analysis->ImmediateDominator(loop_merge_);
+
+  if (!bb) return nullptr;
+
+  if (bb->ctail()->opcode() == SpvOpBranchConditional) {
+    condition_block = bb;
+  }
+
+  return condition_block;
+}
+
+bool Loop::FindNumberOfIterations(const ir::Instruction* induction,
+                                  const ir::Instruction* branch_inst,
+                                  size_t* iterations) const {
+  // From the branch instruction find the branch condition.
+  opt::analysis::DefUseManager* def_use_manager =
+      ir_context_->get_def_use_mgr();
+
+  // Condition instruction from the OpConditionalBranch.
+  ir::Instruction* condition =
+      def_use_manager->GetDef(branch_inst->GetSingleWordOperand(0));
+
+  assert(condition->opcode() == SpvOpSLessThan);
+
+  // The right hand side operand of the operation.
+  const ir::Instruction* rhs_inst =
+      def_use_manager->GetDef(condition->GetSingleWordOperand(3));
+
+  // Exit out if we couldn't resolve the rhs to be a constant integer.
+  uint32_t condition_value = 0;
+  if (!GetConstant(rhs_inst, ir_context_, &condition_value)) return false;
+
+  // Find the instruction which is stepping through the loop.
+  ir::Instruction* step_inst =
+      GetInductionStepOperation(this, induction, ir_context_);
+
+  if (!step_inst) return false;
+
+  // The instruction representing the constant value being applied in the step
+  // operation.
+  const ir::Instruction* step_amount_inst =
+      def_use_manager->GetDef(step_inst->GetSingleWordOperand(3));
+
+  uint32_t step_value = 0;
+  // Exit out if we couldn't resolve the rhs to be a constant integer.
+  if (!GetConstant(step_amount_inst, ir_context_, &step_value)) return false;
+
+  // Find the inital value of the loop and make sure it is a constant integer.
+  uint32_t init_value = 0;
+  if (!GetInductionInitValue(this, induction, ir_context_, &init_value))
+    return false;
+
+  // If iterations is non null then store the value in that.
+  if (iterations) {
+    *iterations = (condition_value / step_value) - init_value;
+  }
+
+  return true;
+}
+
+ir::Instruction* Loop::FindInductionVariable(
```

Comment:
There could be multiple induction variables in a loop.  Is this based on some existing algorithm?  There are several references in the literature that deal with induction variable detection and elimination.  It would be very useful if this was based on one of them.  Easier to document and maintain.

Reply:
Additionally, induction variables could have a positive or negative step.  This code seems to assume and require always monotonically increasing loops.  This should be documented or perhaps it's not too difficult to support?